File: core/src/main/java/heast/core/network/PacketDecoder.java
package heast.core.network;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ReplayingDecoder;

import java.util.List;

public final class PacketDecoder extends ReplayingDecoder<Packet<?>> {

    private final NetworkSide side;

    public PacketDecoder(NetworkSide side) {
        this.side = side;
    }

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
        PacketBuf buffer = new PacketBuf(in);
        int packetId = buffer.readVarInt();

        Packet<?> packet = ctx.channel()
            .attr(ClientConnection.PROTOCOL_KEY).get()
            .getPacketHandler(side)
            .createPacket(packetId, buffer);

        if (packet != null) {
            out.add(packet);
        } else {
            throw new IllegalStateException("Bad packet id: " + packetId);
        }
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        System.err.println("Exception whilst decoding: " + cause.getMessage());
        cause.printStackTrace();
    }
}

File: Core/Network/Codecs/PacketDecoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;

namespace Core.Network.Codecs;

public class PacketDecoder : ReplayingDecoder
{
    public PacketDecoder(NetworkSide side)
    {
        Side = side;
    }

    public NetworkSide Side { get; }

    protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object?> output)
    {
        var buffer = new PacketBuf(input);
        var id = buffer.ReadVarInt();
        var guid = buffer.ReadGuid();

        var packet = context.Channel
            .GetAttribute(ClientConnection.ProtocolKey).Get()
            .GetPacketHandler(Side)
            .CreatePacket(id, buffer);

        if (packet == null)
        {
            throw new InvalidDataException($"Bad packet id ({id})");
        }

        packet.Guid = guid;

        output.Add(packet);
    }

    public override void ExceptionCaught(IChannelHandlerContext context, Exception exception)
    {
        Console.WriteLine($"Exception whilst decoding: {exception.Message}");
    }
}